File: app/src/main/java/com/example/foxizz/navigation/util/ToastUtil.java
package com.example.foxizz.navigation.util;

import android.widget.Toast;

import static com.example.foxizz.navigation.BaseApplication.getContext;

/**
 * Toast工具类
 */
public class ToastUtil {

    private static Toast toast;

    /**
     * 弹出提示信息，默认显示时间为短
     *
     * @param text 提示信息
     */
    public static void showToast(String text) {
        showToast(text, Toast.LENGTH_SHORT);
    }

    /**
     * 弹出提示信息，默认显示时间为短
     *
     * @param resId 提示信息的resId
     */
    public static void showToast(int resId) {
        showToast(resId, Toast.LENGTH_SHORT);
    }

    /**
     * 弹出提示信息
     *
     * @param text     提示信息
     * @param duration 延时（short/long）
     */
    public static void showToast(String text, int duration) {
        if (toast != null) toast.cancel();//销毁上一个
        toast = Toast.makeText(getContext(), null, duration);//弹出下一个
        toast.setText(text);//重设内容，去除小米自带的"appName:"
        toast.show();
    }

    /**
     * 弹出提示信息
     *
     * @param resId    提示信息的resId
     * @param duration 延时（short/long）
     */
    public static void showToast(int resId, int duration) {
        if (toast != null) toast.cancel();//销毁上一个
        toast = Toast.makeText(getContext(), null, duration);//弹出下一个
        toast.setText(resId);//重设内容，去除小米自带的"appName:"
        toast.show();
    }

}

File: base/src/main/java/base/foxizz/util/ToastUtil.kt
package base.foxizz.util

import android.widget.Toast
import androidx.annotation.StringRes
import base.foxizz.BaseApplication.Companion.baseApplication
import base.foxizz.mlh

/**
 * Toast工具类
 */

//全局Toast对象，在下一个Toast弹出时覆盖上一个Toast，不再排队等待
private lateinit var mToast: Toast

//获取新Toast对象
private val newToast get() = Toast.makeText(baseApplication, null, Toast.LENGTH_SHORT)

//设置Toast内容和延时并弹出
private fun showToast(toast: Toast, string: String?, resId: Int?, duration: Int) {
    toast.run {
        string?.let { setText(it) }
        resId?.let { setText(it) }
        this.duration = duration
        show()
    }
}

//确保Toast操作在主线程进行，等待Toast或覆盖Toast
private fun showToast(string: String?, resId: Int?, duration: Int, isWait: Boolean) {
    mlh.post {
        if (isWait) {
            showToast(newToast, string, resId, duration)
        } else {
            if (::mToast.isInitialized) mToast.cancel()
            mToast = newToast
            showToast(mToast, string, resId, duration)
        }
    }
}

/**
 * 弹出提示信息
 *
 * @param string   字符串资源id
 * @param duration 延时（short/long），默认为short
 * @param isWait   是否等待，若该参数设置为true，则新Toast不再会覆盖掉旧Toast，而是会排队等待弹出
 */
fun showToast(string: String, duration: Int = Toast.LENGTH_SHORT, isWait: Boolean = false) =
    if (string.isNotEmpty()) showToast(string, null, duration, isWait) else Unit

fun showToast(string: String, isWait: Boolean) =
    if (string.isNotEmpty()) showToast(string, null, Toast.LENGTH_SHORT, isWait) else Unit

/**
 * 弹出提示信息
 *
 * @param resId    字符串资源id
 * @param duration 延时（short/long），默认为short
 * @param isWait   是否等待，若该参数设置为true，则新Toast不再会覆盖掉旧Toast，而是会排队等待弹出
 */
fun showToast(@StringRes resId: Int, duration: Int = Toast.LENGTH_SHORT, isWait: Boolean = false) =
    showToast(null, resId, duration, isWait)

fun showToast(@StringRes resId: Int, isWait: Boolean) =
    showToast(null, resId, Toast.LENGTH_SHORT, isWait)
